### PR TITLE
feature/eng-3702-add-legacy-output-to-nextroute

### PR DESCRIPTION
**Pull request by nilsbeck**
# Description

Adds legacy output to the legacy-routing template and fixes some additional stuff.

## Changes

* legacy-routing/main.go
* legacy-routing/output.go
* legacy-routing/helper.go
* fleet.go
* fleet_defaults.go

### CODE REVIEW

File: nextroute/schema/fleet.go [hunk 1]
```diff
@@ -121,60 +127,38 @@ type Limits struct {
...
-		}
-		if fleetInput.Defaults.Stops.CompatibilityAttributes != nil {
-			stopCompats = *fleetInput.Defaults.Stops.CompatibilityAttributes
-		}
-		if fleetInput.Defaults.Stops.HardWindow != nil {
-			input.Defaults.Stops.StartTimeWindow = *fleetInput.Defaults.Stops.HardWindow
-		}
+		//Defaults: &Defaults{},
 	}
+	// stopCompats := make([]string, 0)
```
Comment (davidrijsman) [inline nextroute/schema/fleet.go]: Why all this dead code (same applies to all dead code in this file)

File: nextroute/schema/fleet.go [hunk 2]
```diff
@@ -213,48 +245,210 @@ func (fleetInput FleetInput) ToNextRoute() (Input, error) {
...
+	if fleetInput.DurationGroups != nil {
+		input.DurationGroups = &fleetInput.DurationGroups
+	}
 	input.Vehicles = vehicles
 	input.Stops = stops
 
 	return input, nil
 }
 
+// func stopDefaults(fleetInput FleetInput, input Input, stopCompats []string) (Input, []string) {
```
Comment (davidrijsman) [inline nextroute/schema/fleet.go]: dead code

File: templates/routing-legacy/main.go [hunk 3]
```diff
@@ -35,41 +39,149 @@ func solver(
 	ctx context.Context,
 	input schema.FleetInput,
 	options options,
-) (runSchema.Output, error) {
+) (FleetOutput, error) {
 	nextrouteInput, err := input.ToNextRoute()
 	if err != nil {
 		panic(err)
```
Comment (davidrijsman) [inline templates/routing-legacy/main.go]: Why panic and not returning the error?

Comment (nilsbeck) [reply templates/routing-legacy/main.go]: sorry, was not expecting you to look so quickly :D fixed just now with another push.

File: nextroute/schema/fleet.go (hunk 2 above)
Comment (nilsbeck) [reply nextroute/schema/fleet.go]: sorry, was not expecting you to look so quickly :D fixed just now with another push.

File: templates/routing-legacy/main.go [hunk 4]
```diff
@@ -35,41 +39,149 @@ func solver(
...
 	}
 
 	solver, err := nextroute.NewParallelSolver(model)
 	if err != nil {
-		return runSchema.Output{}, err
+		return FleetOutput{}, err
+	}
+
+	solutionsCount := "last"
+	runSolutions, err := run.ParseSolutions(solutionsCount)
```
Comment (davidrijsman) [inline templates/routing-legacy/main.go]: Can't you just do:

`runSolutions := run.Last`

File: nextroute/schema/fleet.go (hunk 1 above)
Comment (nilsbeck) [reply nextroute/schema/fleet.go]: sorry, was not expecting you to look so quickly :D fixed just now with another push.

File: templates/routing-legacy/main.go [hunk 5]
```diff
@@ -35,41 +39,149 @@ func solver(
...
+	}
+
+	solutions, err := solver.Solve(ctx, options.Solve)
+	if err != nil {
+		return FleetOutput{}, err
 	}
 
-	solutions, err := solver.Solve(ctx, solveOptions)
+	// Last solution == 1, all solutions == 0
+	if runSolutions == 1 {
```
Comment (davidrijsman) [inline templates/routing-legacy/main.go]: if runSolutions == run.Last {

Comment (davidrijsman) [reply templates/routing-legacy/main.go]: Which as far as I can see is always the case?

File: templates/routing-legacy/main.go (hunk 4 above)
Comment (nilsbeck) [reply templates/routing-legacy/main.go]: yep. did not see that. thanks

File: templates/routing-legacy/main.go (hunk 5 above)
Comment (nilsbeck) [reply templates/routing-legacy/main.go]: yep.

File: templates/routing-legacy/main.go [hunk 6]
```diff
@@ -35,41 +39,149 @@ func solver(
...
+			output.Statistics.Time.ElapsedSeconds = float64(data.X)
+			output.Statistics.Time.Elapsed = time.Duration(data.X * statistics.Float64(time.Second)).String()
+		}
+	}
 
 	return output, nil
 }
+
+// MapWithError maps a slice of type T to a slice of type R using the function f.
+func MapWithError[T any, R any](v []T, f func(T) (R, error)) ([]R, error) {
```
Comment (davidrijsman) [inline templates/routing-legacy/main.go]: Does not have to be exported (like format)

File: templates/routing-legacy/main.go [hunk 7]
```diff
@@ -35,41 +39,149 @@ func solver(
...
+		solutionArray = append(solutionArray, s)
+	}
+	output, err := format(ctx, options.Solve.Duration, solver, ToFleetSolutionOutput, solutionArray...)
+	if err != nil {
+		return output, err
+	}
+	return output, nil
+}
+
+func format(
```
Comment (davidrijsman) [inline templates/routing-legacy/main.go]: I would define format in output.go

Comment (davidrijsman) [reply templates/routing-legacy/main.go]: I would rewrite the code to something like this:
```
solutionArray := make([]nextroute.Solution, 0)
switch runSolutions {
case run.Last:
  solutionArray = append(solutionArray, solutions.Last())	
case run.All:
  for s := range solutions {
	  solutionArray = append(solutionArray, s)
  }
default:
  return FleetOutput{}, fmt.Errorf("....") 
}
output, err := format(ctx, options.Solve.Duration, solver, ToFleetSolutionOutput, solutionArray...)
if err != nil {
   return output, err
}
return output, nil
```

**Review by davidrijsman — CHANGES_REQUESTED** [review]
See comments

File: templates/routing-legacy/main.go (hunk 7 above)
Comment (nilsbeck) [reply templates/routing-legacy/main.go]: yep, looks a lot better.

File: templates/routing-legacy/main.go (hunk 6 above)
Comment (nilsbeck) [reply templates/routing-legacy/main.go]: changed and moved to `helper.go`

File: templates/routing-legacy/main.go [hunk 8]
```diff
@@ -35,41 +35,61 @@ func solver(
 	ctx context.Context,
 	input schema.FleetInput,
 	options options,
-) (runSchema.Output, error) {
+) (FleetOutput, error) {
 	nextrouteInput, err := input.ToNextRoute()
 	if err != nil {
 		panic(err)
```
Comment (sebastian-quintero) [inline templates/routing-legacy/main.go]: Still has the panic

**Review by sebastian-quintero — APPROVED** [review]

File: templates/routing-legacy/helper.go [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+import (
+	"github.com/nextmv-io/sdk/measure"
+	"github.com/nextmv-io/sdk/nextroute"
+	"github.com/nextmv-io/sdk/nextroute/common"
+	"github.com/nextmv-io/sdk/nextroute/schema"
+	"github.com/nextmv-io/sdk/route"
+	"github.com/twpayne/go-polyline"
+)
+
+func makeDistanceExpression(input schema.FleetInput) nextroute.DistanceExpression {
```
Comment (davidrijsman) [inline templates/routing-legacy/helper.go]: newDistanceExpression, we use the new pre-fix for factory methods / functions

File: templates/routing-legacy/helper.go [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+	return distanceExpression
+}
+
+type distanceData struct {
+	distance nextroute.DistanceExpression
+}
+
+// makeFloatMatrix is a helper function that takes a route.ByIndex and returns a
+// [][]float64.
+func makeFloatMatrix(matrix route.ByIndex, length int) [][]float64 {
```
Comment (davidrijsman) [inline templates/routing-legacy/helper.go]: newFloatMatric, bit of nit picking but same argument as before use new pre-fix for factories

**Review by davidrijsman — APPROVED** [review]
Small non-blocking remarks, if other changes request you could change it but not important enought to be blocking.

**Pull request merged**